examples/esp32/xively_demo/main/xively_if.c

```diff
@@ -0,0 +1,617 @@
+xif_mqtt_topics_t xif_mqtt_topics;
+
+/* libxively Handles */
+static xi_context_handle_t xif_context_handle = -1;
+
+/* FreeRTOS Handles */
+static EventGroupHandle_t xif_requests_event_group_handle = NULL;
+
+#define PUB_INCREASING_COUNTER 0 /* Set 1 to get periodic messages to Button */
+#if PUB_INCREASING_COUNTER
+  #define COUNTER_PUBLISH_PERIOD 10
+  static void xif_publish_counter( void );
+  static xi_timed_task_handle_t xif_pubcounter_scheduled_task_handle = -1;
+#endif /* PUB_INCREASING_COUNTER */
+
+/******************************************************************************
+*                           Xively Interface Implementation
+******************************************************************************/
+void __attribute__((weak)) xif_state_machine_aborted_callback( void )
+{
+    printf( "\n[XIF] The Xively IF state machine and RTOS task are about to shut" );
+    printf( " down due to an unrecoverable error" );
+    printf( "\n[XIF] You should re-implement this function to handle it gracefully" );
+}
+
+int xif_set_device_info( char* xi_acc_id, char* xi_dev_id, char* xi_dev_pwd )
+{
+    xif_mqtt_device_info.xi_account_id = xi_acc_id;
+    xif_mqtt_device_info.xi_device_id  = xi_dev_id;
+    xif_mqtt_device_info.xi_device_pwd = xi_dev_pwd;
+    if( 0 > xif_build_all_mqtt_topics() )
+    {
+        printf( "\n[XIF] Unrecoverable building MQTT topic strings. Abort" );
+        return -1;
+    }
+    return 0;
+}
+
+/*-----------------------------------------------------------------------------
+                           FreeRTOS-specific implementation
+-----------------------------------------------------------------------------*/
+int xif_set_request_bits( xif_action_requests_t requested_action )
+{
+    BaseType_t higher_priority_task_woken, evt_group_set_result;
+
+    /* xHigherPriorityTaskWoken must be initialised to pdFALSE. */
+    higher_priority_task_woken = pdFALSE;
+
+    evt_group_set_result = xEventGroupSetBitsFromISR(
+        xif_requests_event_group_handle, requested_action, &higher_priority_task_woken );
+
+    /* Was the message posted successfully? */
+    if ( evt_group_set_result == pdFAIL )
+    {
+        return -1;
+    }
+    else if( higher_priority_task_woken == pdTRUE )
+    {
+        /* If xHigherPriorityTaskWoken is now set to pdTRUE then a context
+        switch should be requested.  The macro used is port specific and will
+        be either portYIELD_FROM_ISR() or portEND_SWITCHING_ISR() - refer to
+        the documentation page for the port being used. */
+        portYIELD_FROM_ISR();
+    }
+    return 0;
+}
+
+int xif_clear_request_bits( xif_action_requests_t requested_action )
+{
+    BaseType_t result;
+    result =
+        xEventGroupClearBitsFromISR( xif_requests_event_group_handle, requested_action );
+    if ( pdPASS != result )
+    {
+        return -1;
+    }
+    return 0;
+}
+
+int xif_request_action( xif_action_requests_t requested_action )
```

Comment:
I think some comments here are needed. This is for requesting changes in FSM behavior, not for requesting xively actions right?  I would separate the two via nomenclature as to not confuse the reader.

Reply:
Great point :+1: I've added some comments, and I'll find some better nomenclature